src/shmem_internal.h

```diff
@@ -105,11 +121,101 @@ extern int shmem_internal_debug;
         }                                                                                               \
     } while (0)
 
+#define SHMEM_ERR_CHECK_PE(pe)                                          \
+    do {                                                                \
+        if ((pe) < 0 || (pe) >= shmem_internal_num_pes) {               \
+            fprintf(stderr, "ERROR: %s(): PE argument (%d) is invalid\n", \
+                    __func__, (pe));                                    \
+            shmem_runtime_abort(100, PACKAGE_NAME " exited in error");  \
+        }                                                               \
+    } while (0)
+
+#define SHMEM_ERR_CHECK_SYMMETRIC(ptr_in, len)                                          \
+    do {                                                                                \
+        const void *ptr_base = (void*)(ptr_in);                                         \
+        const void *ptr_ext  = (void*)((uint8_t *) (ptr_base) + (len));                 \
+        const void *data_ext = (void*)((uint8_t *) shmem_internal_data_base +           \
+                                                   shmem_internal_data_length);         \
+        const void *heap_ext = (void*)((uint8_t *) shmem_internal_heap_base +           \
+                                                   shmem_internal_heap_length);         \
+        if (len == 0) {                                                                 \
+            break; /* Skip this check when the length is 0 */                           \
+        }                                                                               \
+        else if (ptr_base >= shmem_internal_data_base && ptr_base < data_ext) {         \
+            if (ptr_ext > data_ext) {                                                   \
+                fprintf(stderr, "ERROR: %s(): Argument \"%s\" [%p..%p) exceeds "        \
+                                "sym. data region [%p..%p)\n",                          \
+                        __func__, #ptr_in, ptr_base, ptr_ext,                           \
+                        shmem_internal_data_base, data_ext);                            \
+                shmem_runtime_abort(100, PACKAGE_NAME " exited in error");              \
+            }                                                                           \
+        }                                                                               \
+        else if (ptr_base >= shmem_internal_heap_base && ptr_base < heap_ext) {         \
+            if (ptr_ext > heap_ext) {                                                   \
```

Comment:
we could just combine this into one check right? since if ptr_ext < heap_ext, then it must follow that ptr_base < heap_ext, since by we know ptr_base < ptr_ext